### ISSUE TITLE
CH592

### ISSUE THREAD

**Issue by mejobloggs** [semickolon/kirei#1]
Just curious... Will the CH592 also work? Or does that require different code.

WeAct CH592: https://a.aliexpress.com/_mNbe7W4

I was going to buy the 582 but the 58 2 seems cheaper and newer.

Awesome project btw! I hope to try it out

**Comment by semickolon** [comment]
I'm curious about this too. Looking at their [examples](https://github.com/openwch/ch592), it seems like the code is largely the same. The pre-compiled BLE stack has a different filename ("CH592"), but the API looks the same.

My _guess_ is that it would work by just replacing the BLE stack, startup.S file, and linker. But again, just a guess. I don't have a CH592 board yet to verify this.

P.S. This project wouldn't be possible without your Zig blinky, so thank you!!

**Comment by mejobloggs** [comment]
Wow awesome as. My end goal was to create keyboard firmware too! But I didn't understand how to do this kind of programming, so it was too hard for me 😅

I will be so happy to use your firmware 👍👍

Currently I have a Corne Keyboard, but thinking to design a new keyboard with CH582 or CH592

**Comment by semickolon** [comment]
Awesome, go for it!

I'd just advise not to use B16-23 for row pins because there are no GPIO interrupts there (for less power consumption). For columns, any pin's good.

**Comment by rgoulter** [comment]
> Looking at their [examples](https://github.com/openwch/ch592), it seems like the code is largely the same. The pre-compiled BLE stack has a different filename ("CH592"), but the API looks the same.
>
> My guess is that it would work by just replacing the BLE stack, startup.S file, and linker. But again, just a guess. I don't have a CH592 board yet to verify this.

I had a try at this. https://github.com/rgoulter/kirei/commits/try-ch592/

- I wondered whether the CH58x code would run on the CH592 board. It did not. (The code that runs Kirei on CH582 didn't run Kirei on CH592).

- I'd hoped that replacing 58x-specific stuff would be enough. -- I replaced the `startup.S`, `link.ld`, replaced the `ch58x/lib` headers/libraries from the CH592 EVT; some small adjustments were needed to get things to build.

While I was able to get the LED blinking in a loop on the CH592, the code tripped up when calling `c.GAPRole_PeripheralInit()` in `initPeripheralRole()` in `ble.zig`. -- With a bit of printf debugging, I could observe print statements up until that point, but nothing after.

If I understood why, I'd be able to avoid the problem.

I think maybe the configuration of BLE in the CH592 EVT is slightly different than the EVT that the Kirei CH58x platform ported over.

**Comment by rgoulter** [comment]
Slight progress on this. https://github.com/rgoulter/kirei/commit/e49d0ada8c00cdae62f61c97fb52551207f574e4

I noticed that the CH592 EVT includes `BLE/LIB/ble_task_scheduler.S`. I noticed that the BLE Peripheral example requires this to build, and that the CH58x EVT doesn't include this file.

This relies on a `g_LLE_IRQLibHandlerLocation`, `g_LLE_IRQLibHandlerLocation = (uint32_t)LLE_IRQLibHandler`. (In the EVT, this is implemented in the `CH59x_BLEInit` function in `BLE/HAL/mcu.c`.

The assembly file has `.func`/.endfunc` directives, which are apparently GNU specific.

With that, the BLE peripheral is at least discoverable; although still doesn't successfully connect. (UART1 output shows `ble_dev: GAP state connected`, quickly followed by `disconnected`).

**Comment by rgoulter** [comment]
While trying to narrow down on differences between the kirei code above, and BLE code from the WCH EVT: I noticed that the Peripheral example from the CH592 EVT doesn't work if I change its `config.h` `HAL_SLEEP` to `true`. (More specifically: by default, it's `false`, and I can connect to "Simple Peripheral"; but, by using `HAL_SLEEP`, I'm unable to connect to "Simple Peripheral").

EDIT: The above is true when using the xpack riscv-none-elf-gcc 13. Whereas, when using the riscv-none-embed-gcc 8.2 toolchain distributed by mounriver, `HAL_SLEEP` works as I'd expect with the CH592 EVT code.

I tried seeing if that was a smoking gun for the above; after commenting out `cfg.idleCB = ...` on the kirei code I linked above ... and removing all the BLE/GATT services (battery, device info, HID), then the zig code works as a simple peripheral example.
